stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+    case _Basic_format_arg_type::_Pointer_type:
+        return _STD forward<_Visitor>(_Vis)(_Arg._Pointer_state);
+    case _Basic_format_arg_type::_CString_type:
+        return _STD forward<_Visitor>(_Vis)(_Arg._CString_state);
+    case _Basic_format_arg_type::_String_type:
+        return _STD forward<_Visitor>(_Vis)(_Arg._String_state);
+    case _Basic_format_arg_type::_Custom_type:
+        return _STD forward<_Visitor>(_Vis)(_Arg._Custom_state);
+    default:
+        _STL_VERIFY(false, "basic_format_arg is in impossible state");
+        return _STD forward<_Visitor>(_Vis)(0);
+    }
+}
+
+// we need to implement this ourselves because from_chars does not work with wide characters and isn't constexpr
+template <class _CharT>
+_NODISCARD constexpr const _CharT* _Parse_nonnegative_integer(
+    const _CharT* _Begin, const _CharT* _End, unsigned int& _Value) {
+    _STL_INTERNAL_CHECK(_Begin != _End && '0' <= *_Begin && *_Begin <= '9');
+    _Value                          = 0;
+    constexpr unsigned int _Max_int = static_cast<unsigned int>((numeric_limits<int>::max)());
+    constexpr unsigned int _Big_int = _Max_int / 10;
+
+    do {
+        if (_Value > _Big_int) {
+            _Value = _Max_int + 1;
+            break;
+        }
+        _Value = _Value * 10 + static_cast<unsigned int>(*_Begin - '0');
+        ++_Begin;
+    } while (_Begin != _End && '0' <= *_Begin && *_Begin <= '9');
+    if (_Value > _Max_int) {
+        throw format_error("Number is too big");
+    }
+    return _Begin;
+}
+
+template <class _CharT>
+_NODISCARD constexpr const _CharT* _Parse_nonnegative_integer(const _CharT* _Begin, const _CharT* _End, int& _Value) {
+    unsigned int _Val_unsigned = 0;
+
+    _Begin = _Parse_nonnegative_integer(_Begin, _End, _Val_unsigned);
+    // Never invalid because _Parse_nonnegative_integer throws an error for values that don't fit in signed integers
+    _Value = static_cast<int>(_Val_unsigned);
+    return _Begin;
+}
+
+template <class _CharT, _Parse_arg_id_callbacks<_CharT> _Callbacks_type>
+_NODISCARD constexpr const _CharT* _Parse_arg_id(
+    const _CharT* _Begin, const _CharT* _End, _Callbacks_type&& _Callbacks) {
+    _STL_INTERNAL_CHECK(_Begin != _End);
+    _CharT _Ch = *_Begin;
+    // No id provided, format string is using automatic indexing.
+    if (_Ch == '}' || _Ch == ':') {
+        _Callbacks._On_auto_id();
+        return _Begin;
+    }
+
+    if (_Ch >= '0' && _Ch <= '9') {
+        unsigned int _Index = 0;
+        // arg_id is not allowed to have any leading zeros, but is allowed to be
+        // equal to zero (but not '00'). So if _Ch is zero we skip the parsing, leave
+        // _Index set to zero and let the validity checks below ensure that the arg_id
+        // wasn't something like "00", or "023".
+        if (_Ch != '0') {
+            _Begin = _Parse_nonnegative_integer(_Begin, _End, _Index);
+        } else {
+            ++_Begin;
+        }

```

Comment:
Super nitpick: We usually avoid negations in comparisons also there should be a newline i guess?
```suggestion
        if (_Ch == '0') {
            ++_Begin;
        } else {
            _Begin = _Parse_nonnegative_integer(_Begin, _End, _Index);
        }
        
```